lib/haml_lint/linter/repeated_id.rb

```diff
@@ -0,0 +1,29 @@
+module HamlLint
+  # Detects repeated instances of an element ID in a file
+  class Linter::RepeatedId < Linter
+    include LinterRegistry
+
+    MESSAGE_FORMAT = %{Do not repeat id "#%s" on the page}.freeze
+
+    def visit_tag(node)
+      id = node.tag_id
+      return unless id && !id.empty?
+
+      nodes = (id_map[id] << node)
+      case nodes.size
+      when 2 then nodes.each { |logged_node| add_lint(logged_node, id) }
+      when (3..Float::INFINITY) then add_lint(node, id)
```

Comment:
Yeah, I went back and forth between adding the node to the list first and then checking for repeats and this version. Here's the alternative:

```ruby
def visit_tag(node)
  id = node.tag_id
  return unless id && !id.empty?

  nodes = id_map[i]
  case nodes.size
  when 0 then nodes << node
  when 1 then (nodes << node).each { |logged_node| add_lint(logged_node, id) }
  else nodes << node && add_lint(node, id)
  end
end
```

I went with the version in the PR because I wasn't a fan of the repetition of the `nodes << node`. Doing it up front means we only have to say it once.

Now that I'm thinking about it, I could do this:

```ruby
def visit_tag(node)
  id = node.tag_id
  return unless id && !id.empty?

  nodes = (id_map[id] << node)
  case nodes.size
  when 1 then return
  when 2 then nodes.each { |logged_node| add_lint(logged_node, id) }
  else add_lint(node, id)
  end
end
```

Which could then refactor to the following:

```ruby
def visit_tag(node)
  id = node.tag_id
  return unless id && !id.empty?

  nodes = (id_map[id] << node)
  case nodes.size
  when 1 then return
  when 2 then add_lint_for_first_duplication(nodes)
  else add_lint(node, id)
  end
end

private

def add_lint_for_first_duplication(nodes)
  nodes.each { |node| add_lint(node, node.tag_id) }
end
```

I think that last one is probably the winner. Do you agree?